language/diem-vm/src/diem_transaction_validator.rs

```diff
@@ -150,7 +150,22 @@ pub(crate) fn validate_signature_checked_transaction<R: RemoteCache>(
     let mut cost_strategy =
         CostStrategy::system(vm.get_gas_schedule(&log_context)?, GasUnits::new(0));
     let prologue_status = match transaction.payload() {
-        TransactionPayload::Script(_script) => {
+        TransactionPayload::Script(_) => {
+            vm.check_gas(&txn_data, &log_context)?;
+            vm.run_script_prologue(
+                &mut session,
+                &mut cost_strategy,
+                &txn_data,
+                &currency_code,
+                &log_context,
+            )
+        }
+        TransactionPayload::ScriptFunction(_) => {
+            // gate the behavior until the Diem version is ready
+            if vm.get_diem_version()? < DIEM_VERSION_2 {
+                return Err(VMStatus::Error(StatusCode::FEATURE_UNDER_GATING));
```

Comment:
@runtian-zhou feel free to check whether this is the correct way of gating the behavior until a future `DiemVersion`. Gating in the function `validate_signature_checked_transaction` affects both the validators in the mempool as well as the executors, because the execution flow `execute_user_transaction` also calls this function before casing on the payload type.

Reply:
Which will cause this transaction to be discarded. This seems fine to me. When we have a network with half old nodes and half new nodes and somehow a scriptfun transaction flows through mempool (which it shouldn't due to this gating), the old nodes will simply drop this block due to deserialziation error. Thus this entire block will never be committed. Is that right? cc @zekun000

Reply:
Also what about feature disabling? Should we setup a convention such that version always increments by two and we reserve the odd bit to flip the feature on and off just in case of severe regression in production? cc @aching

Reply:
yes, half/half it won't be committed, but should we gate the execution as well unless the DiemVersion >=2?

imo it's fine to use writeset txn to revert the version instead of bumping by two

Reply:
That's also a great point. Currently DiemVersion can only be incremented in the normal setup. But with WriteSet it's totally doable